File: src/xatu_mcp/tools/execute_python.py
"""Execute Python code in a sandboxed environment."""

from typing import Any

from mcp.server import Server
from mcp.types import TextContent, Tool
import structlog

from xatu_mcp.config import Config
from xatu_mcp.sandbox.base import SandboxBackend

logger = structlog.get_logger()


def register_execute_python(server: Server, sandbox: SandboxBackend, config: Config) -> None:
    """Register the execute_python tool with the MCP server.

    Args:
        server: The MCP server instance.
        sandbox: The sandbox backend to use for execution.
        config: Server configuration.
    """

    @server.list_tools()
    async def list_tools() -> list[Tool]:
        return [
            Tool(
                name="execute_python",
                description="""Execute Python code in a sandboxed environment.

The xatu library is pre-installed for querying Ethereum network data:

```python
from xatu import clickhouse, prometheus, loki, storage

# Query ClickHouse for blockchain data
df = clickhouse.query("mainnet", "SELECT * FROM beacon_api_eth_v1_events_block LIMIT 10")

# Query Prometheus metrics
result = prometheus.query("up")

# Generate and save charts
import matplotlib.pyplot as plt
plt.figure(figsize=(10, 6))
plt.plot(df['slot'], df['block_root'])
plt.savefig('/output/chart.png')

# Upload to get a URL
url = storage.upload('/output/chart.png')
print(f"Chart: {url}")
```

Available ClickHouse clusters:
- "xatu": Production raw data (mainnet, sepolia, holesky, hoodi)
- "xatu-experimental": Devnet raw data
- "xatu-cbt": Aggregated/CBT tables

All output files should be written to /output/ directory.
Data stays in the sandbox - Claude only sees stdout and file URLs.""",
                inputSchema={
                    "type": "object",
                    "properties": {
                        "code": {
                            "type": "string",
                            "description": "Python code to execute",
                        },
                        "timeout": {
                            "type": "integer",
                            "description": "Execution timeout in seconds (default: 60, max: 300)",
                            "minimum": 1,
                            "maximum": 300,
                            "default": 60,
                        },
                    },
                    "required": ["code"],
                },
            ),
        ]

    @server.call_tool()
    async def call_tool(name: str, arguments: dict[str, Any]) -> list[TextContent]:
        if name != "execute_python":
            raise ValueError(f"Unknown tool: {name}")

        code = arguments.get("code")
        if not code:
            raise ValueError("Code is required")

        timeout = arguments.get("timeout", 60)
        timeout = min(max(1, timeout), 300)  # Clamp to 1-300 seconds

        logger.info(
            "Executing Python code",
            code_length=len(code),
            timeout=timeout,
            backend=sandbox.name,
        )

        # Build environment variables for the sandbox
        env = _build_sandbox_env(config)

        try:
            result = await sandbox.execute(code=code, env=env, timeout=timeout)
        except TimeoutError as e:
            logger.warning("Execution timed out", timeout=timeout)
            return [
                TextContent(
                    type="text",
                    text=f"Execution timed out after {timeout} seconds",
                )
            ]
        except Exception as e:
            logger.error("Execution failed", error=str(e))
            return [
                TextContent(
                    type="text",
                    text=f"Execution error: {e}",
                )
            ]

        # Build response
        response_parts = []

        if result.stdout:
            response_parts.append(f"=== STDOUT ===\n{result.stdout}")

        if result.stderr:
            response_parts.append(f"=== STDERR ===\n{result.stderr}")

        if result.output_files:
            files_list = "\n".join(f"  - {f}" for f in result.output_files)
            response_parts.append(f"=== OUTPUT FILES ===\n{files_list}")

        response_parts.append(f"=== EXIT CODE: {result.exit_code} ===")
        response_parts.append(f"=== DURATION: {result.duration_seconds:.2f}s ===")

        logger.info(
            "Execution completed",
            exit_code=result.exit_code,
            duration=result.duration_seconds,
            output_files=result.output_files,
        )

        return [
            TextContent(
                type="text",
                text="\n\n".join(response_parts),
            )
        ]


def _build_sandbox_env(config: Config) -> dict[str, str]:
    """Build environment variables to pass to the sandbox.

    Args:
        config: Server configuration.

    Returns:
        Dictionary of environment variables.
    """
    env: dict[str, str] = {}

    # ClickHouse clusters
    if config.clickhouse.xatu:
        env["XATU_CLICKHOUSE_HOST"] = config.clickhouse.xatu.host
        env["XATU_CLICKHOUSE_PORT"] = str(config.clickhouse.xatu.port)
        env["XATU_CLICKHOUSE_PROTOCOL"] = config.clickhouse.xatu.protocol
        env["XATU_CLICKHOUSE_USER"] = config.clickhouse.xatu.user
        env["XATU_CLICKHOUSE_PASSWORD"] = config.clickhouse.xatu.password
        env["XATU_CLICKHOUSE_DATABASE"] = config.clickhouse.xatu.database

    if config.clickhouse.xatu_experimental:
        env["XATU_EXPERIMENTAL_CLICKHOUSE_HOST"] = config.clickhouse.xatu_experimental.host
        env["XATU_EXPERIMENTAL_CLICKHOUSE_PORT"] = str(config.clickhouse.xatu_experimental.port)
        env["XATU_EXPERIMENTAL_CLICKHOUSE_PROTOCOL"] = config.clickhouse.xatu_experimental.protocol
        env["XATU_EXPERIMENTAL_CLICKHOUSE_USER"] = config.clickhouse.xatu_experimental.user
        env["XATU_EXPERIMENTAL_CLICKHOUSE_PASSWORD"] = config.clickhouse.xatu_experimental.password
        env["XATU_EXPERIMENTAL_CLICKHOUSE_DATABASE"] = config.clickhouse.xatu_experimental.database

    if config.clickhouse.xatu_cbt:
        env["XATU_CBT_CLICKHOUSE_HOST"] = config.clickhouse.xatu_cbt.host
        env["XATU_CBT_CLICKHOUSE_PORT"] = str(config.clickhouse.xatu_cbt.port)
        env["XATU_CBT_CLICKHOUSE_PROTOCOL"] = config.clickhouse.xatu_cbt.protocol
        env["XATU_CBT_CLICKHOUSE_USER"] = config.clickhouse.xatu_cbt.user
        env["XATU_CBT_CLICKHOUSE_PASSWORD"] = config.clickhouse.xatu_cbt.password
        env["XATU_CBT_CLICKHOUSE_DATABASE"] = config.clickhouse.xatu_cbt.database

    # Prometheus
    if config.prometheus:
        env["XATU_PROMETHEUS_URL"] = config.prometheus.url

    # Loki
    if config.loki:
        env["XATU_LOKI_URL"] = config.loki.url

    # S3 Storage
    if config.storage:
        env["XATU_S3_ENDPOINT"] = config.storage.endpoint
        env["XATU_S3_ACCESS_KEY"] = config.storage.access_key
        env["XATU_S3_SECRET_KEY"] = config.storage.secret_key
        env["XATU_S3_BUCKET"] = config.storage.bucket
        env["XATU_S3_REGION"] = config.storage.region
        if config.storage.public_url_prefix:
            env["XATU_S3_PUBLIC_URL_PREFIX"] = config.storage.public_url_prefix

    return env

File: pkg/tool/execute_python.go
package tool

import (
	"context"
	"fmt"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/ethpandaops/xatu-mcp/pkg/auth"
	"github.com/ethpandaops/xatu-mcp/pkg/config"
	"github.com/ethpandaops/xatu-mcp/pkg/sandbox"
	"github.com/mark3labs/mcp-go/mcp"
	"github.com/sirupsen/logrus"
)

// sessionsWithResourceTip tracks sessions that have already seen the resource tip.
var sessionsWithResourceTip sync.Map

// resourceTipMessage is shown after the first execution in a session to guide users to MCP resources.
const resourceTipMessage = `
💡 TIP: Read these MCP resources for available datasources and schemas:
   - datasources://list - available datasource UIDs
   - datasources://clickhouse - ClickHouse datasources only
   - clickhouse-schema://{cluster}/{table} - table schema (if schema discovery enabled)
   - api://xatu - Python library documentation
   - networks://active - available networks`

const (
	// ExecutePythonToolName is the name of the execute_python tool.
	ExecutePythonToolName = "execute_python"
	// DefaultTimeout is the default execution timeout in seconds.
	DefaultTimeout = 60
	// MaxTimeout is the maximum allowed execution timeout in seconds.
	MaxTimeout = 300
	// MinTimeout is the minimum allowed execution timeout in seconds.
	MinTimeout = 1
)

// executePythonDescription is the description of the execute_python tool.
const executePythonDescription = `Execute Python code in a sandboxed environment with the xatu library pre-installed.

Read api://xatu for library documentation. Read datasources://list for available datasource UIDs.

Key modules: clickhouse, prometheus, loki, storage

Files written to /workspace/ persist within a session. Use storage.upload() for public URLs.`

// NewExecutePythonTool creates the execute_python tool definition.
func NewExecutePythonTool(
	log logrus.FieldLogger,
	sandboxSvc sandbox.Service,
	cfg *config.Config,
) Definition {
	return Definition{
		Tool: mcp.Tool{
			Name:        ExecutePythonToolName,
			Description: executePythonDescription,
			InputSchema: mcp.ToolInputSchema{
				Type: "object",
				Properties: map[string]any{
					"code": map[string]any{
						"type":        "string",
						"description": "Python code to execute",
					},
					"timeout": map[string]any{
						"type":        "integer",
						"description": "Execution timeout in seconds (default: from config, max: 300)",
						"minimum":     MinTimeout,
						"maximum":     MaxTimeout,
					},
					"session_id": map[string]any{
						"type":        "string",
						"description": "Optional session ID to reuse a persistent container. If omitted, a new session is created (when sessions are enabled).",
					},
				},
				Required: []string{"code"},
			},
		},
		Handler: newExecutePythonHandler(log, sandboxSvc, cfg),
	}
}

// newExecutePythonHandler creates the handler function for execute_python.
func newExecutePythonHandler(
	log logrus.FieldLogger,
	sandboxSvc sandbox.Service,
	cfg *config.Config,
) Handler {
	handlerLog := log.WithField("tool", ExecutePythonToolName)

	return func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		// Extract code from arguments using mcp-go helper methods.
		code, err := request.RequireString("code")
		if err != nil {
			return CallToolError(fmt.Errorf("invalid arguments: %w", err)), nil
		}

		if code == "" {
			return CallToolError(fmt.Errorf("code is required")), nil
		}

		// Extract timeout from arguments using mcp-go helper with default.
		timeout := request.GetInt("timeout", cfg.Sandbox.Timeout)

		// Validate timeout.
		if timeout < MinTimeout || timeout > MaxTimeout {
			return CallToolError(
				fmt.Errorf("timeout must be between %d and %d seconds", MinTimeout, MaxTimeout),
			), nil
		}

		// Extract optional session_id.
		sessionID := request.GetString("session_id", "")

		// Extract owner ID from auth context for session binding.
		var ownerID string
		if user := auth.GetAuthUser(ctx); user != nil {
			ownerID = fmt.Sprintf("%d", user.GitHubID)
		}

		handlerLog.WithFields(logrus.Fields{
			"code_length": len(code),
			"timeout":     timeout,
			"backend":     sandboxSvc.Name(),
			"session_id":  sessionID,
			"owner_id":    ownerID,
		}).Info("Executing Python code")

		// Build environment variables for the sandbox.
		env := buildSandboxEnv(cfg)

		// Execute the code in the sandbox.
		result, err := sandboxSvc.Execute(ctx, sandbox.ExecuteRequest{
			Code:      code,
			Env:       env,
			Timeout:   time.Duration(timeout) * time.Second,
			SessionID: sessionID,
			OwnerID:   ownerID,
		})
		if err != nil {
			handlerLog.WithError(err).Error("Execution failed")

			return CallToolError(fmt.Errorf("execution error: %w", err)), nil
		}

		// Format the response.
		response := formatExecutionResult(result, cfg)

		// Show resource tip after the first execution in a session.
		sessionKey := result.SessionID
		if sessionKey == "" {
			sessionKey = result.ExecutionID // Use execution ID if no session
		}

		if _, alreadyShown := sessionsWithResourceTip.LoadOrStore(sessionKey, true); !alreadyShown {
			response += resourceTipMessage
		}

		handlerLog.WithFields(logrus.Fields{
			"execution_id": result.ExecutionID,
			"exit_code":    result.ExitCode,
			"duration":     result.DurationSeconds,
			"output_files": result.OutputFiles,
			"session_id":   result.SessionID,
		}).Info("Execution completed")

		return CallToolSuccess(response), nil
	}
}

// formatExecutionResult formats the execution result into a string.
func formatExecutionResult(result *sandbox.ExecutionResult, cfg *config.Config) string {
	var parts []string

	if result.Stdout != "" {
		parts = append(parts, fmt.Sprintf("[stdout]\n%s", result.Stdout))
	}

	if result.Stderr != "" {
		parts = append(parts, fmt.Sprintf("[stderr]\n%s", result.Stderr))
	}

	if len(result.OutputFiles) > 0 {
		parts = append(parts, fmt.Sprintf("[files] %s", strings.Join(result.OutputFiles, ", ")))
	}

	// Include session info if available.
	if result.SessionID != "" {
		sessionInfo := fmt.Sprintf("[session] id=%s ttl=%s",
			result.SessionID, result.SessionTTLRemaining.Round(time.Second))

		if len(result.SessionFiles) > 0 {
			workspaceFiles := make([]string, 0, len(result.SessionFiles))
			for _, f := range result.SessionFiles {
				workspaceFiles = append(workspaceFiles, fmt.Sprintf("%s(%s)", f.Name, formatSize(f.Size)))
			}

			sessionInfo += fmt.Sprintf(" workspace=[%s]", strings.Join(workspaceFiles, ", "))
		}

		parts = append(parts, sessionInfo)
	}

	parts = append(parts, fmt.Sprintf("[exit=%d duration=%.2fs id=%s]",
		result.ExitCode, result.DurationSeconds, result.ExecutionID))

	// Add note about localhost URLs if storage is configured with localhost.
	if cfg.Storage != nil && strings.Contains(cfg.Storage.PublicURLPrefix, "localhost") {
		if strings.Contains(result.Stdout, "localhost") {
			parts = append(parts, "[note] Storage URLs use localhost - these are accessible if running the server locally.")
		}
	}

	return strings.Join(parts, "\n")
}

// formatSize formats a byte size into a human-readable string.
func formatSize(bytes int64) string {
	const unit = 1024

	if bytes < unit {
		return fmt.Sprintf("%d B", bytes)
	}

	div, exp := int64(unit), 0
	for n := bytes / unit; n >= unit; n /= unit {
		div *= unit
		exp++
	}

	return fmt.Sprintf("%.1f %cB", float64(bytes)/float64(div), "KMGTPE"[exp])
}

// buildSandboxEnv creates the environment variables map for the sandbox.
func buildSandboxEnv(cfg *config.Config) map[string]string {
	env := make(map[string]string, 8)

	// Grafana configuration - all datasource queries route through Grafana.
	env["XATU_GRAFANA_URL"] = cfg.Grafana.URL
	env["XATU_GRAFANA_TOKEN"] = cfg.Grafana.ServiceToken
	env["XATU_HTTP_TIMEOUT"] = strconv.Itoa(cfg.Grafana.Timeout)

	// S3 Storage.
	if cfg.Storage != nil {
		env["XATU_S3_ENDPOINT"] = cfg.Storage.Endpoint
		env["XATU_S3_ACCESS_KEY"] = cfg.Storage.AccessKey
		env["XATU_S3_SECRET_KEY"] = cfg.Storage.SecretKey
		env["XATU_S3_BUCKET"] = cfg.Storage.Bucket
		env["XATU_S3_REGION"] = cfg.Storage.Region

		if cfg.Storage.PublicURLPrefix != "" {
			env["XATU_S3_PUBLIC_URL_PREFIX"] = cfg.Storage.PublicURLPrefix
		}
	}

	return env
}
